compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -3186,6 +3186,62 @@ declare_lint! {
     };
 }
 
+declare_lint! {
+    /// The `unstable_syntax_pre_expansion` lint detects the use of unstable
+    /// syntax that is discarded during attribute expansion.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// #[cfg(FALSE)]
+    /// macro foo() {}
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// The input to active attributes such as `#[cfg]` or procedural macro
+    /// attributes is required to be valid syntax. Previously, the compiler only
+    /// gated the use of unstable syntax features after resolving `#[cfg]` gates
+    /// and expanding procedural macros.
+    ///
+    /// To avoid relying on unstable syntax, move the use of unstable syntax
+    /// into a position where the compiler does not parse the syntax, such as a
+    /// functionlike macro.
+    ///
+    /// ```rust
+    /// # #![deny(unstable_syntax_pre_expansion)]
+    ///
+    /// macro_rules identity {
+    ///    ( $($tokens:tt)* ) => { $($tokens) }
+    /// }
+    ///
+    /// #[cfg(FALSE)]
+    /// identity! {
+    ///    macro foo() {}
+    /// }
+    ///
+    /// # #[expect(unstable_syntax_pre_expansion)]
+    /// # identity! {
+    /// #    #[cfg(FALSE)]
+    /// #    macro foo() {}
+    /// # }
```

Comment:
I added this early to make sure this got a warning but now that it's tested in UI tests this can probably be removed?

Reply:
```suggestion
```